### PR TITLE
TiKV Project Security Self-Assessment - Security Pals

### PR DESCRIPTION
**Pull request by Michael-Susu12138**
Created and added first draft for TiKV Project Security Self-Assessment.

Please feel free to share your feedback on the security self-assessment.

### CODE REVIEW

**Comment by eddie-knight** [comment]
Hi there! I'm just getting started looking at your pull request, and I noticed the DCO check is failing.

You can look at the checks section of the PR (I believe it should always be below the last comment) and look for a red X highlighting the failed check. In this case, you can click `Details` for more information about how to get that check passing.
Screenshot 2023-12-08 at 8 35 18 AM

File: assessments/projects/TiKV/self-assessment.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+* Vulnerability Scanning and Patching Process: Regularly scan for vulnerabilities in RocksDB and its dependencies, and establish a quick patching process
+(SBOMs, no detailed checkup system for open-source dependencies).
+
+ **Placement Driver Security**
+ * Distributed Trust and Consensus Mechanism: Implement a distributed trust and consensus mechanism for metadata changes in the Placement Driver.
+ * Backup and Recovery Plans: Establish robust backup and recovery plans for the metadata and transaction logs stored in the Placement Driver (Store two extra redundant copies of meta-data for recovery against data corruption – similar to the Raft Consensus Algorithm).
+
+**Timestamp Oracle Scalability and Security**
+* High-Availability Setup for Timestamp Oracle: Implement a high-availability setup for the Timestamp Oracle with automatic failover capabilities.
+* Rate Limiting and Queue Management: Introduce rate limiting and queue management to handle high demand without compromising security or performance.
```
Comment (eddie-knight) [inline assessments/projects/TiKV/self-assessment.md:360]:
```suggestion
```

Following discussions from the review team, we are asking that Security Pals please remove any recommendations  from the scope of this document. You may optionally communicate recommendations for projects in a more accessible location, such as in a project community channel or GitHub Issue.

File: assessments/projects/TiKV/self-assessment.md [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+### Non-goals
+* Not supposed to be a relational database (No SQL)
+* Does not specifically implement features which compromise performance without clear security benefits
+* Perfect security is not a goal of TiKV; the aim should be to mitigate risk to acceptable levels as perfect security is unattainable
+* Does not address end-user security, such as password manager or client side security, which are outside the database’s control
+* Does not attempt to solve external network security issues that are beyond the scope of TiKV
+
+## Self-assessment use
+
+This self-assessment is created by the TiKV team to perform an internal analysis of the
```
Comment (eddie-knight) [inline assessments/projects/TiKV/self-assessment.md:148]: You mention elsewhere in this document that the TiKV team was not available to participate in this self assessment, so this sentence appears inaccurate. Consider adjusting this sentence and using the topmost paragraph (below the first header) to list the names of all contributors to this document.

File: assessments/projects/TiKV/self-assessment.md [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+* [Questions about potential vulnerabilities](#questions-about-potential-vulnerabilities)
+* [Suggested action items](#suggested-action-items)
+
+## Metadata
+|   |  |
+| -- | -- |
+| Software | [TiKV](https://github.com/tikv/tikv/tree/master)  |
+| Security Provider | No  |
+| Languages | Rust, Java, Go, and C |
+| SBOM |Link to the dependencies file: [Cargo.toml](https://github.com/tikv/tikv/blob/master/Cargo.toml) |
```
Comment (eddie-knight) [inline assessments/projects/TiKV/self-assessment.md:30]:
```suggestion
| SBOM |Link to the dependencies file: TiKV does not currently generate an SBOM on release |
```

The cargo manifest and software bill of materials are different types of items with differing purposes.

File: assessments/projects/TiKV/self-assessment.md (hunk 1 above)
Comment (dominiqueeberhard2) [reply assessments/projects/TiKV/self-assessment.md]: We successfully removed all of the recommendations from our document. Thank you for _your_ recommendation!

File: assessments/projects/TiKV/self-assessment.md (hunk 2 above)
Comment (dominiqueeberhard2) [reply assessments/projects/TiKV/self-assessment.md]: We added our names below the Overview Header and also updated the line you mentioned above (line 148). We wrote that line originally assuming that the TiKV project maintainers would be able to get back to us. Thank you for bringing it to our attention. If our changes do not reflect what you meant in your suggestion, please let us know!

File: PUBLICATIONS.md
Comment (ragashreeshekar) [inline PUBLICATIONS.md:1]: Security pals effort does not require changes to publications document. Please discard the changes to it.

File: README.md
Comment (ragashreeshekar) [inline README.md:1]: Security pals effort does not require changes this document. Please discard the changes to it.

File: assessments/Open_and_Secure.pdf
Comment (ragashreeshekar) [inline assessments/Open_and_Secure.pdf:1]: Security pals effort does not require inserting the Open and secure book. This is already a part of the repo. To keep the PR and repo light, please update the PR to the latest main branch, and remove any file that is not part of security assessment or it's dependent files.

File: website/config.toml
Comment (ragashreeshekar) [inline website/config.toml:1]: Security pals effort does not require changes to website and related documents. Please discard all the changes related to the website.

File: assessments/projects/tikv/self-assessment.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+| Languages | Rust, Java, Go, and C |
+| SBOM |Link to the dependencies file: TiKV does not currently generate an SBOM on release |
+| | |
+
+### Security links
+
+
+| Doc | url |
+| -- | -- |
+| Security file | https://github.com/tikv/tikv/blob/master/security/Security-Audit.pdf |
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:37]: Security file usually refers to security.md which consists of the governance process and procedures in finding and solving security issues. There is one provided by the team, please update https://github.com/tikv/tikv/blob/master/SECURITY.md for the security file and move the security audit report to another row in this table

File: assessments/projects/tikv/self-assessment.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+### Actors
+
+* **Clients**
+* **TiKV Instance**
+  - Data Regions (within TiKV Nodes)
+* **Placement Driver**
+    - TimeStamp Oracle (within Placement Driver)
+
+### Actions
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:113]: Actions refer to what are the steps performed by each actor/multiple actors involved in this system, and the workflow including checks performed in achieving each of these capabilities. Refer https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md for a detailed description. 

Please update this section to reflect the same. A good example for this includes - https://github.com/cncf/tag-security/blob/main/assessments/projects/pixie/self-assessment.md#actions

File: assessments/projects/tikv/self-assessment.md [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+
+
+## Project compliance
+
+Although TiKV does not provide explicit documentation regarding its compliance with specific regulations, it adheres to general best practices in security as an open-source distributed database. Although initially developed for the Chinese market, TiKV's documentation does not explicitly address its compliance with Western regulations like GDPR and HIPAA. The documentation of project compliance can potentially expedite the process of adoption.
+
+## Secure development practices
+
+For testing, TiKV uses regression tests, performance tests, ChaosMesh, and models their algorithms utilizing TLA+. TiKV uses Cargo as a universal project tool, and split tests into test modules in respective code files. Among the fuzzing libraries used are LLVM's libfuzzer, AFL and Google's Honggfuzz; however, tests do not run in an automated fashion.
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:188]: I'm curious to learn the meaning of however, "tests do not run in an automated fashion.". Can you please elaborate.
Do you mean to say there is no CI pipeline that runs these tests? Do they include any form of SAST, SCA or secret scanning service in the development pipeline (manual or automated)?

File: assessments/projects/tikv/self-assessment.md [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+* Decisions regarding the project are made by votes on the primary project community repository.
+* Votes are indicated via pull requests under the votes folder.
+* Positive votes require no explanation, while vetoes (negative votes) must have an explanation. 
+* Required vote counts can vary based on the decision but typically require at least 2 positive votes.
+
+* Internal Communication: Slack 
+
+ * Inbound Communication: Twitter, Blog, Reddit, Stack Overflow
+
+* Outbound Communication: None
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:209]: This section needs further elaboration. Can you please provide link/reference to these communication modes? A good reference is https://github.com/cncf/tag-security/blob/main/assessments/projects/pixie/self-assessment.md#communication-channels

File: assessments/projects/tikv/self-assessment.md [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+
+**Ecosystem**
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:211]: This section is to showcase how the TiKV project aligns within the cloud-native ecosystem. Please update to refer the same. A good example is https://github.com/cncf/tag-security/blob/main/assessments/projects/pixie/self-assessment.md#ecosystem

File: assessments/projects/tikv/self-assessment.md [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+| Elevation of Privilege             | Gaining unauthorized access could lead to elevated privileges within the system. | Accessing accounts with higher privileges could lead to control over critical functions. | An attacker could gain control over data flow and decision-making by becoming a malicious leader. | Altering timestamps could grant unauthorized data access or modifications.    | Exploiting vulnerabilities could lead to unauthorized access or control over data storage. | Accessing elevated privileges through client exploitation can lead to broader system compromise. | Controlling PD could grant significant control over the cluster’s operational aspects. | Manipulating the timestamping process could give undue advantage or access to certain transactions. | Compromised TLS could be exploited to gain access to encrypted communications, leading to privileged information access. | Exploiting rare vulnerabilities in Rust could lead to unauthorized system access. | |
+
+
+
+## Questions about potential vulnerabilities
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:295]: I hope this section was used for understanding the project security posture, however it isn't relevant to document in the security assessment document. Please remove this.

File: assessments/projects/tikv/self-assessment.md [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+* **SST Importer Cleaning Requirement**: The SST Importer needed to clean the SST if no key-value needed to be saved and rewritten.
+* **CPU Usage of Snapshot-Worker Thread**: An enhancement was requested to reduce the CPU usage of this thread to prevent it from blocking scheduling.
+* **Slow Store Check Leader**: This bug potentially affected the ability of a store to advance resolve_ts.
+* **Titan Configuration Tuning**: This was listed as an enhancement issue to optimize the configuration of the Titan engine.
+* **Different Store Size Among Nodes After Enabling Titan**: There were issues with varying store sizes among TiKV nodes after Titan was enabled.
+* **TiKV Out of Memory (OOM)**: There were reported instances of TiKV running out of memory.
+* **Not All TiKV Clients Offer The Same Encryption**: According to TiKV's Security Configuration page, the TiKV Java Client does not currently support TLS. 
+
+
+***CII Best Practices***
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:259]: CII (https://github.com/coreinfrastructure/best-practices-badge) is an initiative that identifies best practices for Free/Libre and Open Source Software (FLOSS) and implements a badging system for those best practices. TiKV is a badge holder - https://www.bestpractices.dev/zh-CN/projects/2574. Please revisit this section to reflect the same. A good reference is https://github.com/cncf/tag-security/blob/main/assessments/projects/pixie/self-assessment.md#cii-best-practices

File: assessments/projects/tikv/self-assessment.md [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+* ***Raw***: A lower-level key-value API for interacting directly with individual key-value pairs
+* ***Transactional***: A higher level key value API that provides ACID semantics
+
+## Security issue resolution
+
+***Note:*** When looking for methods of security issue resolution, there was no easily accessible area to discuss each protocol. This could potentially lead to mishaps like major security vulnerabilities being publicly shown as issues
+
+* Bug Fixes or documentation improvement can be implemented via a normal Github Pull Request workload
+* To contribute “substantial changes”, a RFC (request for comments) document provides a consistent and controlled path for new features to enter the project
+  - RFCs must provide summary, motivation, detailed design, drawbacks, alternative, and unanswered questions
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:227]: Security issues refers to vulnerabilities in the project. This content is not relevant to the security issue resolution. Please remove them

File: assessments/projects/tikv/self-assessment.md [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+  - The Java and Rust clients are considered unstable
+  - The C client is currently in early development
+
+TiKV provides two separate clients: Raw and Transactional
+* ***Raw***: A lower-level key-value API for interacting directly with individual key-value pairs
+* ***Transactional***: A higher level key value API that provides ACID semantics
+
+## Security issue resolution
+
+***Note:*** When looking for methods of security issue resolution, there was no easily accessible area to discuss each protocol. This could potentially lead to mishaps like major security vulnerabilities being publicly shown as issues
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:223]: This is not accurate. The team has the relevant policies in https://github.com/tikv/tikv/blob/master/SECURITY.md. If there are any recommendations to it, please move them to a separate document  within this folder and link it in the appendix section

File: assessments/projects/tikv/self-assessment.md [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+  - Disclosure Policy:
+    - The received security vulnerability report shall be handed over to the security team for follow-up coordination and repair work.
+    - After the vulnerability is confirmed, a draft Security Advisory will be created on Github that lists the details of the vulnerability.
+    - Invite related personnel to discuss the fix.
+    - Fork the temporary private repository on Github, and collaborate to fix the vulnerability.
+    - After the fixed code is merged into all supported versions, the vulnerability will be publicly posted in the GitHub Advisory Database.
+
+## Appendix
+
+***Known Issues Over Time***
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:238]: Most of these seem like security audit or disclosure from the team themselves, thus, outside the scope of security pals efforts. I would suggest providing a short summary with references to find details.

File: assessments/projects/tikv/self-assessment.md [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+***Related Projects / Vendors***
+
+**Company: PingCAP; Offering: TiDB Cloud; Industry: DBaas.**
+PingCAP is the company behind the open-source TiKV project and supplies continuous development of the project. TiDB Cloud is a database-as-a-service (DBaas) provided by PingCAP. It is built using TiKV and TiDB technologies. It offers fully-managed clusters of the open source TiKV database in the cloud. TiDB Cloud manages various operational tasks such as provisioning, upgrades, scaling, monitoring, and ensuring high availability. The main difference is that some users may prefer the convenience of TiDB Cloud as a service to the responsibility of administering their own TiKV environment. Overall, TiDB Cloud is intended for utilization by users who desire to operate it locally or as a service. Both projects give users the opportunity to leverage PingCAP's innovative technology. 
+
+## STRIDE Threat Model
```
Comment (ragashreeshekar) [inline assessments/projects/tikv/self-assessment.md:283]: Please move the threat model to a separate document within this folder and provide a link to it in the appendix

**Review by ragashreeshekar — CHANGES_REQUESTED** [review]
Thanks for the PR @Michael-Susu12138 and team, appreciate the efforts.
I have completed first pass of review and left several comments on section that needs your attention. Please feel free to reach out here or on slack for any questions and clarifications.

Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

File: assessments/Open_and_Secure.pdf
Comment (eddie-knight) [reply assessments/Open_and_Secure.pdf:1]: This is likely due to an oversight while merging the main branch into this repo

Comment (dominiqueeberhard2) [reply assessments/Open_and_Secure.pdf:1]: We accidentally closed our first pull request while trying to resolve the conflicts; However, everything seems to be up and running at the moment. The new pull request number is #1207.

File: assessments/projects/tikv/self-assessment.md (hunk 9 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:294]: You are right; we initially included this section for the project maintainers to see and guide us. However, the project maintainers were not able to get back to us. We have removed this section accordingly. Again, please refer to our new pull request #1207.

File: assessments/projects/tikv/self-assessment.md (hunk 4 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:38]: We have updated the table accordingly. Please refer to our new pull request #1207.

File: assessments/projects/tikv/self-assessment.md (hunk 5 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:112]: We have updated our actions section. We tried our best to follow the example you provided. Once more, please refer to our new pull request #1207. Thank you.

File: assessments/projects/tikv/self-assessment.md (hunk 6 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:187]: We based our initial comment on a remark made in a security audit performed by a security consulting firm called Cure53. We linked to said security audit, which is available in TiKV's project repo. Hopefully, this clears things up.

File: assessments/projects/tikv/self-assessment.md (hunk 7 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:208]: We closely followed the format of the example you provided. Thank you for the guidance!

File: assessments/projects/tikv/self-assessment.md (hunk 8 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:210]: We have rewritten this section and updated it. Please see our new pull request #1207.

File: assessments/projects/tikv/self-assessment.md (hunk 10 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:258]: This section has been updated.

File: assessments/projects/tikv/self-assessment.md (hunk 13 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:237]: We have shortened this section and added relevant links for further exploration of this topic if the reader is interested.

File: assessments/projects/tikv/self-assessment.md (hunk 14 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:282]: We have moved the threat model to a separate file in the folder and provided a link in the appendix as per your suggestions. Thank you.

File: assessments/projects/tikv/self-assessment.md (hunk 11 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:226]: We have updated the section and shortened it according to your feedback. Hopefully, everything is in order now. Thanks!

File: assessments/projects/tikv/self-assessment.md (hunk 12 above)
Comment (dominiqueeberhard2) [reply assessments/projects/tikv/self-assessment.md:222]: We have removed this note and provided a link to the security file in our appendix section. Please refer to our new pull request #1207.

**Pull request closed without merging**